examples/service_discovery.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import aiohttp
+import asyncio
+
+import asab
+import asab.web
+import asab.api
+import asab.zookeeper
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+asab.Config.add_defaults(
+	{
+		"web": {
+			"listen": "0.0.0.0 8089",
+		},
+
+		"zookeeper": {
+			"path": "example",
+			"servers": "localhost:2181"
+		},
+	}
+)
+
+
+class MyApplication(asab.Application):
+
+	async def initialize(self):
+		# Initialize web server
+		self.add_module(asab.web.Module)
+		websvc = self.get_service("asab.WebService")
+		self.WebContainer = asab.web.WebContainer(websvc, "web")
+
+		# Initialize zookeeper
+		self.add_module(asab.zookeeper.Module)
+		self.ZooKeeperService = self.get_service("asab.ZooKeeperService")
+		self.ZooKeeperContainer = asab.zookeeper.ZooKeeperContainer(self.ZooKeeperService, "zookeeper")
+
+		# Service Discovery is part of ApiService and needs its full functionality
+		self.ASABApiService = asab.api.ApiService(self)
+		self.ASABApiService.initialize_web(self.WebContainer)
+		self.ASABApiService.initialize_zookeeper(self.ZooKeeperContainer)
+
+		# Localize Service Discovery Service
+		self.ServiceDiscoveryService = self.get_service("asab.ServiceDiscoveryService")
+
+		self.WebContainer.WebApp.router.add_get('/locate', self.locate_self)
+
+	async def locate_self(self, request):
+		await asyncio.sleep(2)
+		urls = await self.ServiceDiscoveryService.locate(appclass="MyApplication")
+		if urls is None:
+			print("Application was not advertised properly")
+
+		# Get config of the application:
+		config = None
+		async with aiohttp.ClientSession() as session:
+			while len(urls) > 0:  # if more than one url is provided, try them all until one of them returns 200
+				url = "{}/asab/v1/config".format(urls.pop())
+				print("trying to connect to: {}".format(url))
+				async with session.get(url) as resp:
+					if resp.status == 200:
```

Comment:
```
async with asab.api.Discovery.session() as session:
    async with session.get("http+sd://MyApplication.appclass/...") as resp:
        ...
```

You can achieve this with a custom `TCPConnector` and mainly `resolver`.

https://docs.aiohttp.org/en/stable/client_reference.html#tcpconnector

Notice the loop in `_create_direct_connection` method ... the `aiohttp` already iterates over resolved hosts/ports when the connection fails.
That's exactly what we need here. The DiscoveryService need to feed that (thru custom resolver) ip addresses and ports of services to discover.